### PR TITLE
Implement send_message_status payloads for nosip

### PR DESCRIPTION
**Pull request by EricRabil**

### CODE REVIEW

File: database/message.go [hunk 1]
```diff
@@ -59,6 +59,11 @@ func (mq *MessageQuery) GetLastByGUID(chat string, guid string) *Message {
 		"FROM message WHERE chat_guid=$1 AND guid=$2 ORDER BY part DESC LIMIT 1", chat, guid)
 }
 
+func (mq *MessageQuery) GetLastByOnlyGUID(guid string) *Message {
+	return mq.get("SELECT chat_guid, guid, part, mxid, sender_guid, timestamp "+
+		"FROM message WHERE guid=$1 ORDER BY part DESC LIMIT 1", guid)
+}
```
Comment (tulir) [inline database/message.go:65]: This sounds slightly dangerous

File: imessage.go [hunk 2]
```diff
@@ -75,6 +81,39 @@ func (imh *iMessageHandler) HandleMessage(msg *imessage.Message) {
...
 
+func (imh *iMessageHandler) HandleMessageStatus(status *imessage.SendMessageStatus) {
+	msg := imh.bridge.DB.Message.GetLastByOnlyGUID(status.GUID)
+	if msg == nil {
+		imh.log.Infoln("Ignoring message status for unknown message", status.GUID)
+		return
+	}
+	portal := imh.bridge.GetPortalByGUID(msg.ChatGUID)
+	if len(portal.GUID) == 0 {
+		imh.log.Infofln("Ignoring message status for message from unknown portal %s/%s", msg.GUID, msg.ChatGUID)
```
Comment (tulir) [inline imessage.go:92]:
```suggestion
		imh.log.Debugfln("Ignoring message status for message from unknown portal %s/%s", msg.GUID, msg.ChatGUID)
```

File: imessage.go [hunk 3]
```diff
@@ -75,6 +81,39 @@ func (imh *iMessageHandler) HandleMessage(msg *imessage.Message) {
 	portal.Messages <- msg
 }
 
+func (imh *iMessageHandler) HandleMessageStatus(status *imessage.SendMessageStatus) {
+	msg := imh.bridge.DB.Message.GetLastByOnlyGUID(status.GUID)
+	if msg == nil {
+		imh.log.Infoln("Ignoring message status for unknown message", status.GUID)
```
Comment (tulir) [inline imessage.go:87]:
```suggestion
		imh.log.Debugln("Ignoring message status for unknown message", status.GUID)
```

File: imessage/struct.go [hunk 4]
```diff
@@ -242,3 +243,10 @@ type PushKeyRequest struct {
 
 	Data map[string]interface{} `json:"data"`
 }
+
+type SendMessageStatus struct {
+	GUID       string `json:"guid"`
+	Status     string `json:"status"`
+	Message    string `json:"message,omitempty"`
+	StatusCode string `json:"statusCode,omitempty"`
```
Comment (tulir) [inline imessage/struct.go:251]:
```suggestion
	StatusCode string `json:"status_code,omitempty"`
```

**Review by tulir — APPROVED** [review]
Probably fine other than the two log levels and json field name

File: database/message.go (hunk 1 above)
Comment (EricRabil) [reply database/message.go:65]: Why?

**Pull request merged**